File: porject_api/src/main/java/swd392/service/message/MessageService.java
package swd392.service.message;

import swd392.model.dto.State;
import swd392.model.dto.message.MessageDTO;
import swd392.model.entity.message.Message;


public interface MessageService {
    State<Message, String> send(MessageDTO messageDTO);

    MessageDTO toMessageDTO(Message message);
}

File: project_portal/src/app/core/services/message.service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { environment } from '../../../enviroment/enviroment';
import { Message } from '../../model/type/message/message.type';
import { State } from '../../model/type/state.type';

@Injectable({
  providedIn: 'root'
})
export class MessageService {
  http = inject(HttpClient);

  private send$ = new Subject<State<void>>();
  send = this.send$.asObservable();

  handleSend(newMessage: Message): void {
    this.http.post<void>(`${environment.apiUrl}/messages/send`, newMessage)
      .subscribe({
        next: () => this.send$.next(State.Builder().forSuccessEmpty()),
        error: err => this.send$.next(State.Builder().forError(err))
      });
  }
}
